SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -343,13 +365,10 @@ def shl {w : Nat} (x y : IntW w) (flags : NoWrapFlags := {nsw := false , nuw :=
   let y' ← y
   let nsw := flags.nsw
   let nuw := flags.nuw
-  let AddSignedWraps? : Prop := nsw ∧
     -- "If the nsw keyword is present, then the shift produces a poison value if it shifts out any bits that disagree with the resultant sign bit."
-    -- So, if x is positive, we simply have to check that no 1 bit reaches the sign bit after the shift.
-    -- If x is negative we swap every bit (by doing a xor with all ones) and then check the above condition.
-    ((x'.toInt ≥ 0 ∧ (x'.toNat <<< y'.toNat) ≥ 2^(w-1)) ∨ (x'.toInt < 0 ∧ (((BitVec.allOnes w).toNat ^^^ x'.toNat) <<< y'.toNat) ≥ 2^(w-1)))
-  let AddUnsignedWraps? : Prop := nuw ∧ (x'.toNat <<< y'.toNat ≥ 2^w)
-  if (AddSignedWraps? ∨ AddUnsignedWraps?) then
+  let SignedWrap? : Prop := nsw ∧ ((x' <<< y') >>>ₛ y' = x')
+  let UnsignedWrap? : Prop := nuw ∧ ((x' <<< y') >>> y' = x')
+  if (SignedWrap? ∨ UnsignedWrap?) then
```

Comment:
Nice. This looks so much cleaner. Maybe fold the nsw/nuw definitions into the ifs and have two returns for the signed and unsigned case, respectively?